packages/lesswrong/server/resolvers/ultraFeedResolver.ts

```diff
@@ -0,0 +1,571 @@
+): Promise<FeedFullPost[]> {
+  const { currentUser } = context;
+  const recombeeUser = recombeeRequestHelpers.getRecombeeUser(context);
+  let displayPosts: FeedFullPost[] = [];
+
+  if (!recombeeUser) {
+    // eslint-disable-next-line no-console
+    console.warn("UltraFeedResolver: No Recombee user found. Cannot fetch hybrid recommendations.");
+  } else {
+    const settings: HybridRecombeeConfiguration = {
+      hybridScenarios: { fixed: 'hacker-news', configurable: 'recombee-lesswrong-custom' },
+      excludedPostIds: Array.from(servedPostIds),
+      filterSettings: currentUser?.frontpageFilterSettings,
+    };
+
+    try {
+      const recommendedResults = await recombeeApi.getHybridRecommendationsForUser(
+        recombeeUser,
+        limit,
+        settings,
+        context
+      );
+
+      displayPosts = recommendedResults.map((item, idx): FeedFullPost | null => {
+        if (!item.post?._id) return null;
+
+        // Try to determine the scenario - using the same logic hierarchy as RecombeePostsList.tsx
+        let scenario: string | undefined = item.scenario;
+        
+        if (!scenario) {
+          const aboutPostId = aboutPostIdSetting.get();
+          if (aboutPostId && item.post._id === aboutPostId && idx === 0) {
+            scenario = 'welcome-post';
+          } else if (item.curated) {
+            scenario = 'curated';
+          } else if (item.stickied || item.post.sticky) { 
+            scenario = 'stickied';
+          } else if (item.recommId) {
+            if (item.recommId.includes('forum-classic')) {
+              scenario = 'hacker-news';
+            } else if (item.recommId.includes('recombee-lesswrong-custom')) {
+              scenario = 'recombee-lesswrong-custom';
+            }
+          } else {
+            scenario = 'hacker-news';
+          }
+        }
+        
+        const recommInfo = (item.recommId && item.generatedAt) ? {
+          recommId: item.recommId,
+          scenario: scenario || 'unknown',
+          generatedAt: item.generatedAt,
+        } : undefined;
+
+        return {
+          post: item.post,
+          postMetaInfo: {
+            sources: [scenario as FeedItemSourceType],
+            displayStatus: 'expanded',
+            recommInfo: recommInfo,
+          },
+        };
+      }).filter((p): p is FeedFullPost => p !== null);
+
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.error("Error calling getHybridRecommendationsForUser:", error);
+    }
+  }
+  
+  return displayPosts;
+}
+
+interface UltraFeedArgs {
+  limit?: number;
+  cutoff?: Date;
+  offset?: number;
+  sessionId?: string;
+  settings?: string;
+}
+
+type UltraFeedEventInsertData = Pick<DbUltraFeedEvent, 'userId' | 'eventType' | 'collectionName' | 'documentId'>;
+
+/**
+ * UltraFeed resolver
+ */
+export const ultraFeedGraphQLQueries = {
+  UltraFeed: async (_root: void, args: UltraFeedArgs, context: ResolverContext) => {
```

Comment:
This is a 350 line function

Reply:
It might not be a good use of time right now, but in my ideal world it's less than half that (maybe less than a third), having been factored out into smaller functions